File: clients/typescript/src/checkpoint.ts
import { canonical, jsonText } from "./json.js";
import { checkProfile } from "./client.js";
import { decode, encode } from "./wire.js";
import { DASPError, type Event, type JsonObject, type ProfileValidator, type Session } from "./types.js";

export interface Checkpoint {
  session: Session;
  hostSource: string;
  cursor: number;
  state: JsonObject;
  /** Retain evidence with the checkpoint. Missing evidence requires a new trusted view. */
  evidence: Record<string, string>;
}

/** Call only after authenticating and validating the view and selected profile. */
export function checkpointFromView(view: Event<"view">, validateProfile: ProfileValidator): Checkpoint {
  const event = decode(encode(view));
  if (event.type !== "dasp.view.v1") throw new DASPError("checkpoint", "A view is required.");
  checkProfile(validateProfile, event);
  const { session_id, actor_id, profile, cursor, state } = event.data;
  return { session: { session_id, actor_id, profile }, hostSource: event.source, cursor, state, evidence: {} };
}

function identity(event: Event<"update">): string {
  const result: Record<string, unknown> = {};
  for (const key of ["source", "id", "type", "data", "time", "subject", "dataschema"]) {
    if (Object.hasOwn(event, key)) result[key] = event[key];
  }
  return canonical(result);
}

/**
 * Return a new checkpoint only after every update is valid and applied.
 * Save state, cursor, and evidence in one storage transaction before acknowledging delivery.
 */
export function applyUpdates(
  checkpoint: Checkpoint, events: readonly Event<"update">[],
  reduce: (state: JsonObject, event: Event<"update">) => JsonObject,
  validateProfile: ProfileValidator
): Checkpoint {
  if (!Number.isSafeInteger(checkpoint.cursor) || checkpoint.cursor < 0) throw new DASPError("checkpoint", "Invalid applied cursor.");
  const next = structuredClone(checkpoint);
  for (const input of events) {
    const event = decode(encode(input));
    if (event.type !== "dasp.update.v1" || event.source !== next.hostSource || event.data.session_id !== next.session.session_id) {
      throw new DASPError("checkpoint", "Expected an update from the checkpoint session and host.");
    }
    checkProfile(validateProfile, event);
    const sequence = event.data.sequence, proof = identity(event);
    if (sequence <= next.cursor) {
      if (!Object.hasOwn(next.evidence, String(sequence))) throw new DASPError("missing_evidence", "Recover a trusted view before handling this old update.");
      if (next.evidence[sequence] !== proof) throw new DASPError("changed_update", "A saved update changed.");
      continue;
    }
    if (sequence !== next.cursor + 1) throw new DASPError("gap", "Read missing updates before advancing the cursor.");
    const state = reduce(structuredClone(next.state), structuredClone(event));
    if (!state || typeof state !== "object" || Array.isArray(state)) throw new DASPError("checkpoint", "Reducer must return a JSON object.");
    jsonText(state);
    next.state = structuredClone(state);
    next.cursor = sequence;
    next.evidence[sequence] = proof;
  }
  return next;
}

File: clients/elixir/lib/dasp/checkpoint.ex
defmodule DASP.Checkpoint do
  @moduledoc """
  Apply contiguous updates to a saved projection.
  Save the returned state, cursor, and evidence in one transaction before delivery acknowledgment.
  Callbacks must check the selected profile and have no external effects.
  """
  import DASP.Error, only: [fail: 2]
  alias DASP.{Client, Wire}

  @doc "Create a checkpoint from a view whose host context has been authenticated."
  def from_view(view, validate_profile) do
    Wire.protect(fn ->
      event = view |> Wire.encode!() |> Wire.decode!()
      if event["type"] != "dasp.view.v1", do: fail(:checkpoint, "A view is required.")
      Client.profile!(validate_profile, event)
      d = event["data"]

      %{
        "session" => Map.take(d, ["session_id", "actor_id", "profile"]),
        "host_source" => event["source"],
        "cursor" => d["cursor"],
        "state" => d["state"],
        "evidence" => %{}
      }
    end)
  end

  @doc "Return a new checkpoint only if all updates pass validation and the reducer succeeds."
  def apply_updates(checkpoint, events, reduce, validate_profile) do
    Wire.protect(fn ->
      if not is_integer(checkpoint["cursor"]) or checkpoint["cursor"] < 0 or
           checkpoint["cursor"] > 9_007_199_254_740_991,
         do: fail(:checkpoint, "Invalid applied cursor.")

      Enum.reduce(events, checkpoint, fn input, current ->
        event = input |> Wire.encode!() |> Wire.decode!()
        d = event["data"]

        if event["type"] != "dasp.update.v1" or event["source"] != current["host_source"] or
             d["session_id"] != current["session"]["session_id"],
           do: fail(:checkpoint, "Expected an update from the checkpoint session and host.")

        Client.profile!(validate_profile, event)
        sequence = d["sequence"]
        key = Integer.to_string(sequence)
        proof = Map.take(event, ["source", "id", "type", "data", "time", "subject", "dataschema"])

        cond do
          sequence <= current["cursor"] ->
            if not Map.has_key?(current["evidence"], key),
              do:
                fail(:missing_evidence, "Recover a trusted view before handling this old update.")

            if current["evidence"][key] != proof,
              do: fail(:changed_update, "A saved update changed.")

            current

          sequence != current["cursor"] + 1 ->
            fail(:gap, "Read missing updates before advancing the cursor.")

          true ->
            state = reduce.(current["state"], event)

            if not is_map(state) or is_struct(state),
              do: fail(:checkpoint, "Reducer must return a JSON object.")

            DASP.JSON.encode!(state)

            %{
              current
              | "state" => state,
                "cursor" => sequence,
                "evidence" => Map.put(current["evidence"], key, proof)
            }
        end
      end)
    end)
  end
end
